neqo-transport/src/connection.rs

```diff
@@ -1208,26 +1286,36 @@ impl Connection {
         // Frames for different epochs must go in different packets, but then these
         // packets can go in a single datagram
         for epoch in 0..NUM_EPOCHS {
+            let space = PNSpace::from(epoch);
             let mut encoder = Encoder::default();
             let mut ds = Vec::new();
             let mut tokens = Vec::new();
-            // Try to make our own crypo state and if we can't, skip this
-            // epoch.
-            if self.obtain_crypto_state(epoch).is_err() {
-                continue;
-            }
 
-            if let Some(recvd) = self
-                .obtain_crypto_state(epoch)
-                .as_mut()
-                .unwrap()
-                .recvd_state()
+            // Try to make our own crypo state and if we can't, skip this epoch.
             {
-                let acks = recvd.get_eligible_ack_ranges();
-                Frame::encode_ack_frame(&acks, &mut encoder);
-                // TODO(ekr@rtfm.com): Deal with the case where ACKs don't fit
-                // in an entire packet.
-                assert!(encoder.len() <= self.pmtu);
+                let cs = match self.obtain_crypto_state(epoch) {
+                    Ok(c) => c,
+                    _ => continue,
+                };
+                if cs.tx.is_none() {
+                    continue;
+                }
+            }
+
+            // Always send an ACK if there is one to send.
+            if self.acks[space].ack_now(cur_time) {
+                let mut recvd = mem::replace(&mut self.acks[space], RecvdPackets::new(space));
+                if let Some((frame, token)) = recvd.generate(
```

Comment:
This looks like it wants to go on the list of generators, but I'm guessing something prevents this? If so, could you open an issue and describe what's blocking it?